File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/infrastructure/InMemoryGiftCardEventBus.java
package it.giannibombelli.wsc2026.giftcard.infrastructure;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.application.events.EventBus;
import it.giannibombelli.wsc2026.common.application.events.EventSubscriber;
import it.giannibombelli.wsc2026.giftcard.domain.events.GiftCardEvent;

import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.Executor;

public final class InMemoryGiftCardEventBus implements EventBus<GiftCardEvent> {

    private final Map<Class<? extends GiftCardEvent>, List<EventSubscriber<? extends GiftCardEvent>>> subscribers = new ConcurrentHashMap<>();
    private final Executor executor;

    public InMemoryGiftCardEventBus(Executor executor) {
        Require.requireDependency(executor, "executor");
        this.executor = executor;
    }

    @Override
    public void publish(GiftCardEvent event) {
        Require.requireArgument(event, "event");

        List<EventSubscriber<? extends GiftCardEvent>> matching = subscribers.entrySet().stream()
            .filter(entry -> entry.getKey().isAssignableFrom(event.getClass()))
            .flatMap(entry -> entry.getValue().stream())
            .toList();

        executor.execute(() -> matching.forEach(subscriber -> notify(subscriber, event)));
    }

    @SuppressWarnings("unchecked")
    private <E extends GiftCardEvent> void notify(EventSubscriber<E> subscriber, GiftCardEvent event) {
        subscriber.on((E) event);
    }

    @Override
    public void subscribe(Class<? extends GiftCardEvent> eventType, EventSubscriber<? extends GiftCardEvent> subscriber) {
        Require.requireArgument(eventType, "eventType");
        Require.requireArgument(subscriber, "subscriber");

        subscribers.computeIfAbsent(eventType, k -> new CopyOnWriteArrayList<>()).add(subscriber);
    }
}

File: typescript/src/giftcard/infrastructure/inMemoryGiftCardEventBus.ts
// Event bus in-memory specializzato per GiftCardEvent.
// Supporta dispatch sincrono nei test e configurabile in produzione tramite Executor.

import type { EventBus } from '@/common/application/events/eventBus.js';
import type { EventSubscriber } from '@/common/application/events/eventSubscriber.js';
import type { GiftCardEvent } from '../domain/events/giftCardEvent.js';
import { requireArgument } from '@/common/utils/requireArgument.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export type Executor = (task: () => void) => void;

export class InMemoryGiftCardEventBus implements EventBus<GiftCardEvent> {
  private readonly _subscribers = new Map<string, EventSubscriber<GiftCardEvent>[]>();
  private readonly _executor: Executor;

  constructor(executor: Executor) {
    requireDependency(executor, "executor");
    this._executor = executor;
  }

  publish(event: GiftCardEvent): void {
    requireArgument(event, 'event');

    const list = this._subscribers.get(event.kind) ?? [];
    this._executor(() => list.forEach((subscriber) => subscriber.on(event)));
  }

  subscribe(eventKind: string, subscriber: EventSubscriber<GiftCardEvent>): void {
    requireArgument(eventKind, 'eventKind');
    requireArgument(subscriber, 'subscriber');
    const list = this._subscribers.get(eventKind) ?? [];
    list.push(subscriber);
    this._subscribers.set(eventKind, list);
  }
}
